source/sessions/tests/README.rst

```diff
@@ -0,0 +1,71 @@
+====================
+Driver Session Tests
+====================
+
+.. contents::
+
+----
+
+Introduction
+============
+
+The YAML and JSON files in this directory are platform-independent tests that
+drivers can use to prove their conformance to the Driver Sessions Spec. They are
+designed with the intention of sharing most test-runner code with the
+Transactions spec tests.
+
+Several prose tests, which are not easily expressed in YAML, are also presented
+in the Driver Sessions Spec. Those tests will need to be manually implemented
+by each driver.
+
+Test Format
+===========
+
+The same as the Transactions spec tests format.
+
+Special Test Operations
+```````````````````````
+
+Certain operations that appear in the "operations" array do not correspond to
+API methods but instead represent special test operations. Such operations are
+defined on the "testRunner" object and are documented in the Transactions spec
+test. Additional, session test specific operations are documented here:
+
+assertServerSessionPoolSize
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The "assertServerSessionPoolSize" operation instructs the test runner to
+assert that the test's MongoClient has a ServerSession pool containing the
+given number of ServerSessions::
+
+      - name: assertServerSessionPoolSize
```

Comment:
If session pools were configurable, I suppose we could run tests with a pool size of one and assert that subsequent commands didn't reuse the same LSID (proving that a dirty session was not reused). But session pools are not configurable, so this seemingly requires inspection of the driver's internal state.

I'm not a fan of introducing spec test operations that require internal APIs, as it makes this difficult if not impossible to implement in wrapping drivers (e.g. anything using libmongoc). Would this be the first case of a spec test requiring this (outside of perhaps the URI tests)?

That said, I wouldn't veto this change if it's necessary.

Reply:
Okay, I suspected that this would would hard or impossible to implement in various drivers. Do you have any suggestions for alternative tests?

Perhaps I should revive [this integration test](https://github.com/mongodb/specifications/pull/544/commits/cacd9a2438de793a92b1bf81c55411c2cb5c8d90#diff-1d90a5641b222162674af2ff11f1fea6L13) which indirectly tests that drivers discard dirty sessions. (I removed it because of a race condition in 3.6 sharded clusters)

Reply:
How about a test like this:

- set failpoint that closes the connection
- execute op 1
- execute op 2
- assert that the commandStartedEvent for op1 and op2 have different lsid field values

Reply:
Like this?:

```
    failPoint:
        configureFailPoint: failCommand
        mode: { times: 1 }
        data:
            failCommands: ["insert"]
            closeConnection: true
    operations:
      - name: insertOne
        object: collection
        arguments:
          document: {_id: 2}
        result:
          insertedId: 2
      - name: insertOne
        object: collection
        arguments:
          document: {_id: 3}
        result:
          insertedId: 3
      - name: assertDifferentLsidOnLastTwoCommands
        object: testRunner
```

Reply:
That test seems reasonable. Provided no other thread is using the client, I believe we can rely on the LIFO design of the implicit session pool.

If this is one test, I'd advocate for prose rather than introduce a `assertDifferentLsidOnLastTwoCommands` operation. A custom operation that depends on analyzing captured APM events for previous commands in the same `operations` array seems like it could be a headache to implement in some test runners that segregate the `operations` and `expectations` blocks.

Reply:
Why would it be a headache to allow the `operations` block to also have access to APM events?

Personally, I think very specialized testRunner assertions are much more maintainable than add prose tests. In my experience, prose tests take longer to implement and MUCH longer to code review because the reviewer also needs to read each prose test in the spec and make sure each test is properly translated into code. This continues to be a pain when new prose tests are added and old prose tests are updated.

For example, consider how easy it is to review my implementation of this spec change in python: https://github.com/ShaneHarvey/mongo-python-driver/blob/SPEC-1159/test/test_session.py#L1304-L1331
And compare that to the 1000+ lines of prose tests in the same file.

Reply:
> Why would it be a headache to allow the operations block to also have access to APM events?

I had libmongoc C's test runner in mind when I wrote that, but it looks like even though the APM monitoring and operation execution are defined in two different components, the array where observed events are appended is still accessible during operation execution.

Feel free to ignore my second paragraph in https://github.com/mongodb/specifications/pull/544#discussion_r283812069.